server/automute.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"strconv"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	NotifyPropAutomuted = "msteams_automute"
+
+	// This preference tracks if we've enabled automuting for the user by muting all of their connected channels.
+	PreferenceNameAutomuteEnabled = "automute_enabled"
+)
+
+func (p *Plugin) tryEnableAutomute(userID string, skipConnectedCheck, skipPrimaryPlatformCheck bool) (bool, error) {
+	enabled, err := p.shouldEnableAutomuteForUser(userID, skipConnectedCheck, skipPrimaryPlatformCheck)
+	if err != nil {
+		return false, err
+	} else if !enabled {
+		return false, nil
+	}
+
+	return p.setAutomuteEnabledForUser(userID, true)
+}
+
+// shouldEnableAutomuteForUser returns true if the given user is both connected to Teams and has their primary platform
+// set to Teams.
+//
+// skipConnectedCheck and skipPrimaryPlatformCheck can be used to skip the respective checks if the calling code has
+// already performed those checks.
+func (p *Plugin) shouldEnableAutomuteForUser(userID string, skipConnectedCheck, skipPrimaryPlatformCheck bool) (bool, error) {
+	var connected bool
+	if skipConnectedCheck {
+		connected = true
+	} else {
+		var err error
+		connected, err = p.isUserConnected(userID)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	var teamsPrimary bool
+	if skipPrimaryPlatformCheck {
+		teamsPrimary = true
+	} else {
+		teamsPrimary = p.isUsersPrimaryPlatformTeams(userID)
+	}
+
+	if !connected || !teamsPrimary {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (p *Plugin) tryDisableAutomute(userID string) (bool, error) {
+	return p.setAutomuteEnabledForUser(userID, false)
+}
+
+func (p *Plugin) setAutomuteEnabledForUser(userID string, automuteEnabled bool) (bool, error) {
+	if channelsAutomuted := p.getAutomuteIsEnabledForUser(userID); channelsAutomuted == automuteEnabled {
+		// We've already automuted all the users' channels, so there's nothing to do
+		return false, nil
+	}
+
+	var membersToMute []*model.ChannelMemberIdentifier
+
+	if channels, appErr := p.API.GetChannelsForTeamForUser("", userID, true); appErr != nil {
+		return false, errors.Wrap(appErr, fmt.Sprintf("Unable to get channels for user %s to automute them", userID))
+	} else {
+		for _, channel := range channels {
+			if linked, err := p.canAutomuteChannel(channel); err != nil {
+				return false, err
```

Comment:
Do we want to return potentially half way through? may make sense, but double checking.

Reply:
I'd propose we log the errors, but keep working through it. If it did ever fail, is there anything that would trigger this to "catch up"?

Reply:
The only catch up logic right now would be for the user to toggle their primary platform to disable and re-enable automuting which could run into the same error here